hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestHSync.java

```diff
@@ -279,18 +249,69 @@ public void testOverwriteHSyncFile() throws Exception {
     }
   }
 
-  static void runTestHSync(FileSystem fs, Path file, int initialDataSize)
```

Comment:
The runTestHSync() used by testO3fsHSync() and testOfsHSync() are meant to verify the behavior of concurrent write and read of the same file. It's also the behavior required by HBase.

IMO we should modify runTestHSync() but we should need to verify this behavior.

Reply:
The ITestOzoneContractCreate and ITestRootedOzoneContractCreate failed because the current KeyInputStream reads the data until the file length, while the file length is calculated by the block length which is updated by hsync. 

I checked the HDFS behavior.  DFSInputStream will query DN for the last block length if the file's last block is UC. That's why HDFS can pass the FS contract test.  

After we reduce the hsync call to OM, we should also implement such read logic in KeyInputStream too. Let's see what I can do.